### ISSUE TITLE
Websocket issues

### ISSUE THREAD

**Issue by Vertux** [movim/movim#797]
I just did a fresh git install of movim, which worked so far without any issues and I can run the daemon.
But when I opened the admin web GUI I ran into an websocket issue, the browser cannot establish the conntection to the movim service. The browser console show these errors:

![bildschirmfoto 2019-02-16 um 13 19 06](https://user-images.githubusercontent.com/10457918/52899986-ee36ed80-31f0-11e9-9d12-e043343c772d.png)

The weird thing is, that I cannot see any error in the Apache logs.
I am running Apache/2.4.25, PHP 7.2.15 FPM, postgresql v9.6.10, Debian 9.7

I have no idea where to look further.

**Comment by edhelas** [comment]
Can you check if the cache directory is writable?

**Comment by Vertux** [comment]
@edhelas Yes, the cache directory is writeable by the movim user, It already contains many files. 
I run into this issue with Apache and Nignx, so I think the problem is triggered by FPM. Unfortunately I have no meaningfull error message in the logs.

btw, the server is managed with Plesk 17.8.11, just for reference. I have a hunch that Plesk is responsible for this strange problem.

P.S.: Meanwhile I did a test installation in a VM with a similar environment and it worked out of the box.

**Comment by edhelas** [comment]
Can you please check the other logs than Apache? Like in syslog or log/errors.log ?

**Comment by Vertux** [comment]
Sorry for answering so late, but I had a busy week.
The only movim related error I can find in the syslog is:

`movim[17253]: movim.ERROR: Constant SQL_DATE already defined in /var/www/xxxxxxx/xxxxxx.de/xxxxxxxx/movim/src/Movim/Bootstrap.php (line 105)  [] []`

![bildschirmfoto 2019-03-02 um 13 04 06](https://user-images.githubusercontent.com/10457918/53681576-d892ef00-3ceb-11e9-9dbe-4f533d579a85.png)

If I check the source of the above syntax error I get:

> An error occured, check syslog for more information
> var BASE_HOST = 'movim.xxxxxx.de';
> var BASE_URI = '//movim.xxxxxx.de/';
> var ERROR_URI = '//movim.xxxxxx.de/?disconnect';
> var SECURE_WEBSOCKET = true;
> var SMALL_PICTURE_LIMIT = 512000;

Meanwhile I think the websocket variable is not set correctly somewhere, maybe by Plesk.
I think I am gonna delete the movim subdomain and do a completely new installation this weekend.

**EDIT:** I did a new installation, everything ran fine, but the problem stays the same, I am clueless.

**Comment by Vertux** [comment]
After I turned on debug logging for the mod_proxy I have found this:

```
[proxy:debug] proxy_util.c(1779): AH00925: initializing worker ws://localhost:8080/ shared
[proxy:debug] proxy_util.c(1821): AH00927: initializing worker ws://localhost:8080/ local
[proxy:debug] proxy_util.c(1856): AH00930: initialized pool in child 15408 for (localhost) min=0 max=25 smax=25
[proxy:trace2] mod_proxy.c(656): [client xxxxx] AH03461: attempting to match URI path '/' against prefix '/ws/' for proxying
[proxy:trace2] mod_proxy.c(656): [client xxxxx] AH03461: attempting to match URI path '/index.php' against prefix '/ws/' for proxying
[proxy:trace2] proxy_util.c(1985): [client xxxxx] *: using default reverse proxy worker for unix:///var/www/vhosts/system/movim/php-fpm.sock|fcgi://127.0.$
[proxy:trace2] proxy_util.c(1941): [client xxxxx] *: rewrite of url due to UDS(/var/www/vhosts/system/movim/php-fpm.sock): fcgi://127.0.0.1:9000/var/www/v$
[proxy:debug] mod_proxy.c(1228): [client xxxxx] AH01143: Running scheme unix handler (attempt 0)
[proxy:debug] proxy_util.c(2156): AH00942: FCGI: has acquired connection for (*)
[proxy:debug] proxy_util.c(2209): [client 93.195.140.78:34112] AH00944: connecting fcgi://127.0.0.1:9000/var/www/vhosts/xxxxxx/xxxxxx/movim/index.php to 127.0.0.1:9000
[proxy:debug] proxy_util.c(2246): [client 93.195.140.78:34112] AH02545: fcgi: has determined UDS as /var/www/vhosts/system/movim/php-fpm.sock
[proxy:debug] proxy_util.c(2418): [client 93.195.140.78:34112] AH00947: connected /var/www/vhosts/xxxxx/xxxxx/movim/index.php to httpd-UDS:0
[proxy:debug] proxy_util.c(2786): AH02823: FCGI: connection established with Unix domain socket /var/www/vhosts/system/movim/php-fpm.sock (*)

```
Unfortunately I have no idea what I have to modify in the configuration to make it work. I guess I have to overwrite one or more default prarameters which were set by Plesk.

My working Movim test VM shows these log entries:

```
[proxy:trace2] [pid 9188] mod_proxy.c(656): [client 192.168.20.51:60514] AH03461: attempting to match URI path '/ws/' against prefix '/ws/' for proxying
[proxy:trace1] [pid 9188] mod_proxy.c(741): [client 192.168.20.51:60514] AH03464: URI path '/ws/' matches proxy handler 'proxy:ws://localhost:8080/'
[proxy:trace2] [pid 9188] proxy_util.c(1962): [client 192.168.20.51:60514] ws: found worker ws://localhost:8080/ for ws://localhost:8080/?path=admin&offset=-60
[proxy:debug] [pid 9188] mod_proxy.c(1228): [client 192.168.20.51:60514] AH01143: Running scheme ws handler (attempt 0)
[proxy:debug] [pid 9188] proxy_util.c(2156): AH00942: WS: has acquired connection for (localhost)
[proxy:debug] [pid 9188] proxy_util.c(2209): [client 192.168.20.51:60514] AH00944: connecting ws://localhost:8080/?path=admin&offset=-60 to localhost:8080
[proxy:debug] [pid 9188] proxy_util.c(2418): [client 192.168.20.51:60514] AH00947: connected /?path=admin&offset=-60 to localhost:8080
[proxy:trace2] [pid 9188] proxy_util.c(2850): WS: fam 10 socket created to connect to localhost
```

**Comment by poVoq** [comment]
In the Plesk web host I have the Apache is proxied behind a Nginx, and I was told this results in websocket issues. Not sure if possible with Movim, but maybe there is an option to run it directly through the Nginx?

**Comment by Vertux** [comment]
@poVoq In my case I am running Movim with Apache, Nginx is completely disabled. I have already tried to run Movim with Nginx (Apache disabled), but it does not work too.

**Comment by edhelas** [comment]
@Vertux is your issue somehow related to https://github.com/movim/movim/issues/820 ?

**Comment by Vertux** [comment]
> @Vertux is your issue somehow related to #820 ?

I have the suspicion that the problems are related. The main issue of #797 seems to be this one:

![socket_error](https://user-images.githubusercontent.com/10457918/56428245-faa5f800-62be-11e9-956f-5360359eb05a.png)

The automatic repair process fails. It does not help to delete the socket either, because if I restart movim it recreates the socket and I have the same problem again. Meanwhile it become pretty frustrating.

**Comment by edhelas** [comment]
Hi, did you found the solution ?

**Comment by Vertux** [comment]
@edhelas Unfortunately not, I am gonna close the issue for now, thank you for your support so far.

**Comment by Vertux** [comment]
@edhelas I still struggeling with this issue. One difference is, that I use PHP 7.3.13 at the moment. The apache congig looks like that:
```
ProxyPass /ws/ ws://localhost:8080/
Alias / /var/xxxxxxxxx/mov/public/
<Directory /var/xxxxxxxxxx/mov/public/>
	DirectoryIndex index.php
	Options +FollowSymLinks -Indexes
	AllowOverride FileInfo Options
</Directory>

<Location />
	Header set Access-Control-Allow-Origin "*"
</Location>
```
I get the login page and the admin page. But one thing I am wonder about is, that the server connection is very fragile. I see the page and after a couple of seconds the connection gets lost, only if I reload the page I can reconnect, thats a pretty odd behaviour.
If I login on the admn page I always see this:
![Bildschirmfoto 2019-12-21 um 12 36 06](https://user-images.githubusercontent.com/10457918/71307555-9b215e80-23f0-11ea-93f5-449d2a808c14.png)
Is it possible that this issue is triggered by a namespace problem? I am absolutely sure, that it has something to do with the Plesk server management environment, I guess Movim expects something, which is different in this environment, but I have no idea what it is.

I wish someone more expirenced than me with a Plesk managed server could check it.

**Comment by Vertux** [comment]
Now I have found the reason for these websocket issues. In the [default Plesk configuration](https://docs.plesk.com/en-US/obsidian/administrator-guide/web-servers/apache-and-nginx-web-servers-linux/apache-with-nginx.70837/), nginx is running as reverse proxy server and even if I disable nginx for a subdomain, it still handles the web traffic for that subdomain. As long as no websockets are involved everything works fine. I have now turned off nginx and Movim works perfectly as expected.  
If I have time I am gonna try to figure out what I have to change in the nginx proxy configuration to make it handle websockets correctly.
